lib/sax.js

```diff
@@ -538,86 +538,228 @@ function _copy(source, target) {
+	}
+
+	function compare(text) {
+		return source.substring(index, index + text.length) === text;
+	}
+
+	function getMatch(args) {
+		var expr = grammar.reg('^(', args, ')');
+		var match = source.substring(index).match(expr);
+		if (match) {
+			skip(match[1].length);
+			return match[1];
+		}
+		return null;
+	}
+
+	function parsePI() {
+		var match = source.substring(index).match(grammar.PI);
+		if (!match) {
+			return errorHandler.fatalError('Invalid processing instruction');
+		}
+		if (index > 0 && match[1].toLowerCase() === 'xml') {
+			return errorHandler.fatalError('xml declaration is allowed only at the start of the document');
+		}
+		skip(match[0].length);
+		return match[0];
+	}
+
+	switch (char(2)) {
 		case '-':
-			if (source.charAt(start + 3) === '-') {
-				var end = source.indexOf('-->', start + 4);
-				//append comment source.substring(4,end)//<!--
-				if (end > start) {
-					domBuilder.comment(source, start + 4, end - start - 4);
-					return end + 3;
-				} else {
-					errorHandler.error('Unclosed comment');
-					return -1;
-				}
+			// should be a comment
+			var comment = getMatch(grammar.Comment);
+			if (comment) {
+				domBuilder.comment(comment, 4, comment.length - 7);
+				return index;
 			} else {
-				//error
-				return -1;
+				return errorHandler.fatalError('Invalid comment starting at position ' + start);
 			}
-		default:
-			if (source.substr(start + 3, 6) == 'CDATA[') {
-				var end = source.indexOf(']]>', start + 9);
+		case '[':
+			// should be CDATA
+			var cdata = getMatch(grammar.CDSect);
+			if (cdata) {
 				domBuilder.startCDATA();
-				domBuilder.characters(source, start + 9, end - start - 9);
+				domBuilder.characters(cdata, 9, cdata.length - 12);
 				domBuilder.endCDATA();
-				return end + 3;
+				return index;
+			} else {
+				return errorHandler.fatalError('Invalid CDATA starting at position ' + start);
 			}
-			//<!DOCTYPE
-			//startDTD(java.lang.String name, java.lang.String publicId, java.lang.String systemId)
-			var matchs = split(source, start);
-			var len = matchs.length;
-			if (len > 1 && /!doctype/i.test(matchs[0][0])) {
-				var name = matchs[1][0];
-				var pubid = false;
-				var sysid = false;
-				if (len > 3) {
-					if (/^public$/i.test(matchs[2][0])) {
-						pubid = matchs[3][0];
-						sysid = len > 4 && matchs[4][0];
-					} else if (/^system$/i.test(matchs[2][0])) {
-						sysid = matchs[3][0];
-					}
+		case 'D': {
+			// should be DOCTYPE
+			var doctype = {
+				name: undefined,
+				publicId: undefined,
+				systemId: undefined,
+				internalSubset: undefined,
+			};
+
+			if (!compare('<!DOCTYPE')) {
+				return errorHandler.fatalError('Expected <!DOCTYPE');
+			}
+			skip(9);
+			if (skipBlanks() < 1) {
+				return errorHandler.fatalError('No blanks between DOCTYPE and document name');
+			}
+
+			// Parse the DOCTYPE name
+			doctype.name = getMatch(grammar.Name);
+			if (!doctype.name) return errorHandler.fatalError('DOCTYPE without name');
```

Comment:
PUBLIC and SYSTEM would match Name, if the name is left out, should we check for name matching those?

Reply:
WDYM? Name is required. And i think PUBLIC and SYSTEM are valid names.

```
<!DOCTYPE SYSTEM>
<SYSTEM/>
```

Reply:
interesting, I was coming from the perspective that a doctype couldwrongly be written as 
`<!DOCTYPE SYSTEM "">` or `<!DOCTYPE PUBLIC "" "">` which would not be well formed, and that we might be able to detect that earlier, but you are right.
`<!DOCTYPE SYSTEM SYSTEM "">` would be well formed, so it doesn't make sense.